src/openrct2-ui/windows/StaffList.cpp

```diff
@@ -112,648 +78,625 @@ static rct_widget window_staff_list_widgets[] = {
+        }
     }
 
-    int32_t staffTypeStringId = StaffNamingConvention[selectedTab].plural;
-    // If the number of staff for a given type is 1, we use the singular forms of the names
-    if (StaffList.size() == 1)
+    void RefreshList()
     {
-        staffTypeStringId = StaffNamingConvention[selectedTab].singular;
-    }
+        _staffList.clear();
 
-    auto ft = Formatter();
-    ft.Add<uint16_t>(StaffList.size());
-    ft.Add<rct_string_id>(staffTypeStringId);
-
-    gfx_draw_string_left(
-        dpi, STR_STAFF_LIST_COUNTER, ft.Data(), COLOUR_BLACK,
-        w->windowPos + ScreenCoordsXY{ 4, window_staff_list_widgets[WIDX_STAFF_LIST_LIST].bottom + 2 });
-}
+        for (auto peep : EntityList<Staff>(EntityListId::Peep))
+        {
+            sprite_set_flashing(peep, false);
+            if (peep->AssignedStaffType == GetSelectedStaffType())
+            {
+                sprite_set_flashing(peep, true);
+                _staffList.push_back(peep->sprite_index);
+            }
+        }
 
-/** rct2: 0x00992A08 */
-static constexpr const uint32_t staffOrderBaseSprites[] = {
-    SPR_STAFF_ORDERS_SWEEPING,
-    SPR_STAFF_ORDERS_INSPECT_RIDES,
-    0,
-    0,
-};
+        std::sort(
+            _staffList.begin(), _staffList.end(), [](const uint16_t a, const uint16_t b) { return peep_compare(a, b) < 0; });
+    }
 
-static constexpr const uint32_t staffCostumeSprites[] = {
-    SPR_STAFF_COSTUME_PANDA,   SPR_STAFF_COSTUME_TIGER,   SPR_STAFF_COSTUME_ELEPHANT, SPR_STAFF_COSTUME_ROMAN,
-    SPR_STAFF_COSTUME_GORILLA, SPR_STAFF_COSTUME_SNOWMAN, SPR_STAFF_COSTUME_KNIGHT,   SPR_STAFF_COSTUME_ASTRONAUT,
-    SPR_STAFF_COSTUME_BANDIT,  SPR_STAFF_COSTUME_SHERIFF, SPR_STAFF_COSTUME_PIRATE,
-};
+private:
+    StaffType GetSelectedStaffType() const
+    {
+        return static_cast<StaffType>(_selectedTab);
+    }
 
-/**
- *
- *  rct2: 0x006BD785
- */
-void window_staff_list_scrollpaint(rct_window* w, rct_drawpixelinfo* dpi, int32_t scrollIndex)
-{
-    auto dpiCoords = ScreenCoordsXY{ dpi->x, dpi->y };
-    gfx_fill_rect(
-        dpi, { dpiCoords, dpiCoords + ScreenCoordsXY{ dpi->width - 1, dpi->height - 1 } }, ColourMapA[w->colours[1]].mid_light);
-
-    // How much space do we have for the name and action columns? (Discount scroll area and icons.)
-    const int32_t nonIconSpace = w->widgets[WIDX_STAFF_LIST_LIST].width() - 15 - 68;
-    const int32_t nameColumnSize = nonIconSpace * 0.42;
-    const int32_t actionColumnSize = nonIconSpace * 0.58;
-    const int32_t actionOffset = w->widgets[WIDX_STAFF_LIST_LIST].right - actionColumnSize - 15;
-
-    auto y = 0;
-    auto i = 0;
-    for (auto spriteIndex : StaffList)
+    void DrawTabImages(rct_drawpixelinfo& dpi)
     {
-        if (y > dpi->y + dpi->height)
+        // Handymen tab image
+        auto selectedTab = _selectedTab;
+        auto imageId = (selectedTab == 0 ? (_tabAnimationIndex & ~3) : 0);
+        imageId += GetPeepAnimation(PeepSpriteType::Handyman).base_image + 1;
```

Comment:
I feel the tab drawing should be simplified into a loop but maybe for some later time.